csharp_dotnetcore/12.NLP-With-LUIS/LuisBot/AppInsights/MyAppInsightLuisRecognizer.cs

```diff
@@ -0,0 +1,124 @@
+        /// </summary>
+        /// <param name="context">Context object containing information for a single turn of conversation with a user.</param>
+        /// <param name="ct">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <param name="logOriginalMessage">Determines if the original message is logged into Application Insights.  This is a privacy consideration.</param>
+        /// <returns>The LUIS results of the analysis of the current message text in the current turn's context activity.</returns>
+        public async Task<RecognizerResult> RecognizeAsync(ITurnContext context, CancellationToken ct, bool logOriginalMessage = false)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // Call Luis Recognizer
+            var recognizerResult = await base.RecognizeAsync(context, ct);
+
+            var conversationId = context.Activity.Conversation.Id;
+
+            // Find the Telemetry Client
+            if (context.TurnState.TryGetValue(MyAppInsightsLoggerMiddleware.AppInsightsServiceKey, out var telemetryClient) && recognizerResult != null)
+            {
+                var topLuisIntent = recognizerResult.GetTopScoringIntent();
+                var intentScore = topLuisIntent.score.ToString("N2");
+
+                // Add the intent score and conversation id properties
+                var telemetryProperties = new Dictionary<string, string>()
+                {
+                    { MyLuisConstants.ActivityIdProperty, context.Activity.Id },
+                    { MyLuisConstants.IntentProperty, topLuisIntent.intent },
+                    { MyLuisConstants.IntentScoreProperty, intentScore },
+                };
+
+                if (recognizerResult.Properties.TryGetValue("sentiment", out var sentiment) && sentiment is JObject)
+                {
+                    if (((JObject)sentiment).TryGetValue("label", out var label))
+                    {
+                        telemetryProperties.Add(MyLuisConstants.SentimentLabelProperty, label.Value<string>());
+                    }
+
+                    if (((JObject)sentiment).TryGetValue("score", out var score))
+                    {
+                        telemetryProperties.Add(MyLuisConstants.SentimentScoreProperty, score.Value<string>());
+                    }
+                }
+
+                if (!string.IsNullOrEmpty(conversationId))
+                {
+                    telemetryProperties.Add(MyLuisConstants.ConversationIdProperty, conversationId);
+                }
+
+                // Add Luis Entitites
+                var entities = new List<string>();
+                foreach (var entity in recognizerResult.Entities)
+                {
+                    if (!entity.Key.ToString().Equals("$instance"))
```

Comment:
I would consider adding a link to the LUIS docs for this. Reading this code, I went "huh? What is this?" and had to go read...